src/CloudNative.CloudEvents.SystemTextJson/JsonEventFormatter.cs

```diff
@@ -120,31 +120,72 @@ public JsonEventFormatter(JsonSerializerOptions serializerOptions, JsonDocumentO
+            DecodeBatchModeMessageImpl(new MemoryStream(data), contentType, extensionAttributes, false).GetAwaiter().GetResult();
+
+        private async Task<IReadOnlyList<CloudEvent>> DecodeBatchModeMessageImpl(Stream data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes, bool async)
+        {
+            data = data ?? throw new ArgumentNullException(nameof(data));
+            var document = await ReadDocumentAsync(data, contentType, async).ConfigureAwait(false);
+            using (document)
+            {
+                var root = document.RootElement;
+                if (root.ValueKind != JsonValueKind.Array)
+                {
+                    throw new ArgumentException($"Cannot decode JSON element of kind '{root.ValueKind}' as batch CloudEvent");
+                }
+                // Avoiding LINQ to avoid extraneous allocations etc.
+                List<CloudEvent> events = new List<CloudEvent>(root.GetArrayLength());
+                foreach (var element in root.EnumerateArray())
+                {
+                    events.Add(DecodeJsonElement(element, extensionAttributes));
+                }
+                return events;
+            }
+        }
+
+        private async Task<JsonDocument> ReadDocumentAsync(Stream data, ContentType contentType, bool async)
+        {
             var encoding = contentType.GetEncoding();
-            JsonDocument document;
             if (encoding is UTF8Encoding)
             {
-                document = async
+                return async
                     ? await JsonDocument.ParseAsync(data, DocumentOptions).ConfigureAwait(false)
                     : JsonDocument.Parse(data, DocumentOptions);
             }
             else
             {
                 using var reader = new StreamReader(data, encoding);
-                string json = async
+                var json = async
                     ? await reader.ReadToEndAsync().ConfigureAwait(false)
                     : reader.ReadToEnd();
-                document = JsonDocument.Parse(json, DocumentOptions);
-            }
-            using (document)
-            {
-                return DecodeJsonDocument(document, extensionAttributes);
+                return JsonDocument.Parse(json, DocumentOptions);
             }
         }
 
-        private CloudEvent DecodeJsonDocument(JsonDocument document, IEnumerable<CloudEventAttribute> extensionAttributes = null)
+        // TODO: Override the other methods
+
+        private CloudEvent DecodeJsonElement(JsonElement element, IEnumerable<CloudEventAttribute> extensionAttributes = null)
         {
-            if (!document.RootElement.TryGetProperty(CloudEventsSpecVersion.SpecVersionAttribute.Name, out var specVersionProperty)
+            if (element.ValueKind != JsonValueKind.Object)
+            {
+                throw new ArgumentException($"Cannot decode JSON element of kind '{element.ValueKind}' as CloudEvent");
```

Comment:
Might be worth including the element itself as well.

Reply:
As per earlier comment, I'd generally prefer to avoid including arbitrary data in the exception message. I could be persuaded though - let's chat about it.